File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/TransactionStore.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.service.store.database.StoreTransactionRunner
import com.gemwallet.android.data.service.store.database.TransactionsDao
import com.gemwallet.android.data.service.store.database.entities.toRecord
import com.gemwallet.android.data.repositories.transactions.addSwapMetadata
import com.gemwallet.android.serializer.decodeJson
import com.wallet.core.primitives.Transaction
import com.wallet.core.primitives.WalletId
import uniffi.gemstone.GemTransactionStore

class GemstoneTransactionStore(
    private val transactionsDao: TransactionsDao,
    private val transactionRunner: StoreTransactionRunner,
) : GemTransactionStore {
    override suspend fun saveTransactions(walletId: String, transactions: List<String>) {
        val records = transactions.map { it.decodeJson<Transaction>() }
        transactionRunner.run {
            transactionsDao.insert(records.toRecord(WalletId(walletId)))
            transactionsDao.addSwapMetadata(records)
        }
    }
}

File: ios/Packages/GemstoneServices/Sources/Stores/TransactionStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.AssetId
import typealias Gemstone.Transaction
import protocol Gemstone.GemTransactionStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneTransactionStore: GemTransactionStore, @unchecked Sendable {
    private let store: TransactionStore

    public init(store: TransactionStore) {
        self.store = store
    }

    public func saveTransactions(walletId: String, transactions: [Gemstone.Transaction]) async throws {
        try store.addTransactions(walletId: WalletId.from(id: walletId), transactions: transactions.map { try Primitives.Transaction($0) })
    }
}
